CMakeLists.txt

```diff
@@ -323,19 +329,22 @@ set (MONGOCRYPT_BUILD_VERSION ${BUILD_VERSION})
 configure_file (src/mongocrypt.h.in src/mongocrypt.h)
 
 set (PROJECT_VERSION "${BUILD_VERSION}")
-# If libbson is statically linked, don't require a dependency.
+set (PROJECT_DESCRIPTION "The libmongocrypt client-side field level encryption library.")
+set (PKG_CONFIG_REQUIRES "libbson-static-1.0")
+set (PKG_CONFIG_STATIC_REQUIRES "libbson-static-1.0")
```

Comment:
Ah, I think this was my initial understanding of static/dynamic dependencies for the four cases of linking to libmongocrypt.

1. static link libmongocrypt, and libmongocrypt static links libbson
2. static link libmongocrypt, and libmongocrypt dynamic links libbson
3. dynamic link libmongocrypt, and libmongocrypt static links libbson
4. dynamic link libmongocrypt, and libmongocrypt dynamic links libbson

For (1), a consumer only needs to link to libmongocrypt-static.a.
For (2), a consumer needs to link to libmongocrypt-static.a and libbson.so (since static libraries don't have relocation info).
For (3) and (4), a consumer only needs to link to libmongocrypt.so.

But given the report and this change, that seems incorrect. If libmongocrypt static links libbson, consumers that are static linking libmongocrypt still need to static link libbson?

Reply:
Your (1) is incorrect.  Regardless of whether libmongocrypt was told to dynamically or statically link libbson, consumers of libmongocrypt will also need to specify linkage to libbson (even if the consumer does not directly use libbson).  A static library is nothing more than a collection of object files in a single archive.  It knows nothing of how to link to any other libraries, thus requiring any artifact that links to that static library to also link to anything else that the static library may need.  That is really the fundamental reason why MONGOCRYPT-277 happened in the first place; the pkg-config script did not specify the other things needed by the static libmongocrypt so that consumers of static libmongocrypt could tell the linker about them.

Your (2) is right, but for the wrong reason.  The comments above apply here as well; it is not the absence of relocation info, but rather the absence of linkage info.

For (3) and (4), you are correct.

Reply:
Ah gotcha, thanks for the explanation! For some reason I had thought if libmongocrypt.a was built statically linking to libbson.a, it'd be pulling in libbson's object files.

Reply:
No problem.  Actually, if you think back to when we merged the kms_message project into libmonogcrypt to form a single project it was precisely because prior to the merger we would have to ship libmongocrypt.a and libkms_message.a and the consumer would have to specify both when linking.  By combining the projects, we can produce a libmongocrypt.a that contains all of the libmongocrypt object files and all of the kms_message object files as well.  That is the only way to eliminate the need for the consumer to specify linkage to an upstream dependency of a static library.